### PR TITLE
Form support

### PR DESCRIPTION
**Pull request by johanjanssens**
This PR requires: https://github.com/joomlatools/joomlatools-framework/pull/299

## Form support

This PR implements form support for Pages.  The form is defined through the frontmatter. 

````yaml
form:
    name: contact
    fields:
        name: [string, required]
        email: [email, required]
        subject: [string, required]
        message: [string, required]
   processors: [csv, database: {table: name}, email: {recipient: info@foo.com, title: Contact Form, subject: New submission for contact form}]
    redirect: thank-you
    honeypot: name_01234567
````

The form definition includes the form fields validation rules, form processors, page to redirect to after a successful submission,  and enables the honeypot. The Form API also offers support for POST only forms, aka form endpoints.

_Note_: The form API only works for POST forms, GET forms are not supported. A form can only have a single endpoint, while a page can contain multiple forms if they use different endpoints.

### Name

The name of the form is used by the processors, a single processor can handle different forms and can differentiate between them using the name. If no name is defined in the form the page slug will be used instead. 

### Fields

The fields define the list of accepted fields and the validation rules for each field. Only the fields that are defined will be accepted, any not defined fields that are passed as part of the request payload will be discarded. 

`field: [filter, [required]]`

A field definition is composed as follows, the name of the field, followed by an array of filters. Filters are applied in FIFO order.  For example:

`email: [email, required]`

This will filter the email field as email and will ensure the field is submitted. 

If any of the filter rules fail a `ComPagesControllerExceptionRequestInvalid` exception will be throw which results in a `400 Bad Request` http error.

### Processors

The form processors can process the validated form data. Processors are called in FIFO order.  Out of the box two processors are provided:

#### csv

The csv processor will store the submitted form data in a csv file, each new submission will be a new line added to the file. The default file will be stored at /joomlatools-pages/logs/[name].

The location can be changed be setting the `path` property. For example:

````yaml
form:
    processors: [csv: {path: /path/to/directory}]
````

#### email

The email processor will send the submitted form data using a basic html formatted email. It offers following properties:

- recipients: [array], allows to define one or more recipients for the email (required)
- email: [email], the email of the sender 
- name: [string], the name of the sender
- title: [string], allows to define the message title 
- subject: [string], allows to define the email subject

Only the  recipients property is required, the email and name are based on the Joomla email configuration settings and the title and subject created based on the form name.  You can easily set them as follows:

````yaml
form:
    processors: [email: {email: info@foo.bar, name: Foobar, title: Foobar form, subject: New Foobar Form Submission}]
````

#### database

The database processor will store the submissions in a database table. 

The table needs to be configured by setting the `table` property. For example:

````yaml
form:
    processors: [database: {table: contact_submissions}] 
````

The table property also accepts a fully qualified database table identifier, and can be extended from `ComPagesDatabaseTableAbstract` It’s advised to put them in a /extensions/database/table subfolder. The identifier would then be `ext:database.table.[name]`

#### custom

Custom processors can be added by extending the `ComPagesControllerProcessorAbstract` class. It’s advised to put them in a /extensions/processor subfolder. The identifier would then be `ext:processor.[name]`

### Redirect

The redirect defines the page or the url to redirect to after a successful form submission. Redirects are not triggered if an error happened during form processing.  A url needs to be fully qualified.

### Honeypot

The honeypot offers protection against spam bots. The majority of these bots are pretty dumb and can easily be deceived adding an invisible field to your form that should never contain a value when submitted. Such a field is called a honeypot. These spam bots will just fill all fields, including the honeypot.

If enabled the `honeypot` form property defines the name of the hidden field that is included in the form. For best results we recommend using a descriptive name and appending it with a random part of example `name_[random]`. 

A honeypot needs to be included in your form by using the the `form.honeypot` template helper. For example:

````php
<?= helper('form.honeypot', [honeypot]); ?>
````
Following html will be inserted into the form

````html
<div style="display:none !important;">
   <input type="text" name="'[honeypot]'" value="" autocomplete="false" tabindex="-1">
</div>
````
When a submission comes in with a filled honeypot field a `ComPagesControllerExceptionRequestBlocked` exception will be throw will results in a`403 Forbidden` http error.

### Form only endpoint

If a form doesn’t include a layout and page body the form will be POST only, the url endpoint will not be accessible through a GET request, doing so will result in a `405 Method Not Allowed` error. 

To create a form endpoint simply create a page and only include the form definition in the frontmatter, do **NOT** define a layout and a page body. For example

````yaml
---
form:
    name: […]
    processors: […]
    fields: […]
    redirect: […]
    honeypot: […]
---
````
## Additional improvements

- This PR adds a isForm() method to the page entity, this can be used determine if a page is a form or not. For example:

````
<? if(page('contact')->isForm()); ?>
….
<? endif ?>
````

- This PR adds two new exceptions `ComPagesControllerExceptionRequestBlocked` 
 and `ComPagesControllerExceptionRequestInvalid and are throw if form validation fails.

Both implement the `ComPagesControllerExceptionSpam` interface to make it easy to catch and log spam attempts. 

- This PR improves the returned http methods in the allow header. For none form pages only GET, HEAD and OPTIONS are returned, for forms  GET, POST, HEAD and OPTIONS, or only POST in case of a form endpoint.

### CODE REVIEW

**Comment by johanjanssens** [comment]
@ercanozkaya All ready for review.

File: code/site/components/com_pages/controller/form.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        parent::_initialize($config);
+    }
+
+    protected function _actionSubmit(KControllerContextInterface $context)
+    {
+        $page = $this->getModel()->getPage();
+
+        //Validate the request
+        $this->setHoneypot($page->form->honeypot);
+        $this->setValidationtRules((array) KObjectConfig::unbox($page->form->fields));
```
Comment (ercanozkaya) [inline code/site/components/com_pages/controller/form.php]:
```suggestion
        $this->setValidationRules((array) KObjectConfig::unbox($page->form->fields));
```

File: code/site/components/com_pages/controller/behavior/validatable.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        $this->__honeypot = $name;
+        return $this;
+    }
+
+    public function getHoneypot()
+    {
+        return $this->__honeypot;
+    }
+
+    public function setValidationtRules(array $rules)
```
Comment (ercanozkaya) [inline code/site/components/com_pages/controller/behavior/validatable.php]:
```suggestion
    public function setValidationRules(array $rules)
```

**Comment by johanjanssens** [comment]
> @johanjanssens Left you two typo fixes as suggestions.
> 
> Also, what do you think of making _action=submit the default in either the dispatcher or the controller for POST requests? This would save us from adding it to the form.

Perfect! That solves it, was looking how to do this best, wanted a submit action not add, and didn’t consider overriding the dispatcher. All done now.

**Pull request merged**